src/components/text-editor/__internal__/text-editor.component.js

```diff
@@ -0,0 +1,256 @@
+        setActiveInlines({ BOLD: hasInlineStyle(value, 'BOLD'), ITALIC: hasInlineStyle(value, 'ITALIC') });
+
+        return true;
+      }
+    }
+    onChange(value);
+
+    return false;
+  }
+
+  function handlePastedText(pastedText) {
+    const selectedTextLength = getSelectedLength(value);
+    const newLength = (contentLength + pastedText.length) - selectedTextLength;
+
+    // prevent paste event when the new content would exceed the limit
+    if (newLength > characterLimit) {
+      return 'handled';
+    }
+    setActiveInlines({});
+
+    return 'not-handled';
+  }
+
+  const getEditorState = () => {
+    let editorState = getDecoratedValue(value);
+
+    // should the cursor position be forced to the end of the content
+    if (contentLength > 0 && moveCursor.current && isFocused) {
+      editorState = moveSelectionToEnd(editorState);
+      moveCursor.current = false;
+    }
+
+    return editorState;
+  };
+
+  const editorState = getEditorState();
+
+  const activeControls = {
+    BOLD: activeInlines.BOLD !== undefined ? activeInlines.BOLD : hasInlineStyle(editorState, INLINE_STYLES[0]),
+    ITALIC: activeInlines.ITALIC !== undefined ? activeInlines.ITALIC : hasInlineStyle(editorState, INLINE_STYLES[1]),
+    'unordered-list-item': hasBlockStyle(editorState, BLOCK_TYPES[0]),
+    'ordered-list-item': hasBlockStyle(editorState, BLOCK_TYPES[1])
+  };
+
+  const handleEditorFocus = useCallback((focusValue) => {
+    moveCursor.current = true;
+    setIsFocused(focusValue);
+
+    if (!isFocused && focusValue && editor.current !== document.activeElement) {
+      editor.current.focus();
+    }
+  }, [editor, isFocused]);
+
+  const handleInlineStyleChange = (ev, style) => {
+    ev.preventDefault();
+    setActiveInlines({ ...activeInlines, [style]: !hasInlineStyle(value, style) });
+    handleEditorFocus(true);
+    setInlines([...inlines, style]);
+  };
+
+  const handleBlockStyleChange = (ev, blockType) => {
+    ev.preventDefault();
+    handleEditorFocus(true);
+    onChange(RichUtils.toggleBlockType(value, blockType));
+    const temp = [];
+    INLINE_STYLES.forEach((style) => {
+      if (activeInlines[style] !== undefined) {
+        temp.push(style);
+      }
+    });
+    setInlines(temp);
+  };
+
+  useEffect(() => {
+    // apply the inline styling, having it run in as an effect ensures that styles can be added
+    // even when the editor is not focused
+    INLINE_STYLES.forEach((style) => {
+      const preserveStyle = activeInlines[style] !== undefined && activeInlines[style] !== hasInlineStyle(value, style);
+
+      if (preserveStyle || (isFocused && inlines.includes(style))) {
+        onChange(RichUtils.toggleInlineStyle(value, style));
+        setInlines(inlines.filter(inline => inline !== style));
+      }
+    });
+  }, [activeInlines, contentLength, inlines, isFocused, onChange, value]);
+
+  const { labelId } = label.props;
+
+  return (
+    <>
+      { label }
```

Comment:
Is there a reason to use composition here? Is it possible that `label` won't be a `Label` component in some cases?

Reply:
That's a fair point, the component relies on `labelId` so composition is unnecessary here